File: nodejs-tools/nodejs-compiler-frontend/lexer-xml/src/readIdentifier.ts
/* eslint-disable @typescript-eslint/no-use-before-define */
import { ALPHABET, DIGITS, not } from '@vighnesh153/utils';
import { XmlLexer } from './Lexer';

export function readIdentifier(lexer: XmlLexer): string {
  /* v8 ignore next 3 */
  if (not(isAcceptableIdentifierCharacter(lexer.inputReader.currentCharacter))) {
    throw new Error(`You should not attempt to read an identifier which doesn't start with '_' or a letter`);
  }

  const identifierBuilder: string[] = [];
  while (isAcceptableIdentifierCharacter(lexer.inputReader.peekCharacter())) {
    identifierBuilder.push(lexer.inputReader.currentCharacter!);
    lexer.inputReader.readNextCharacter();
  }
  if (isAcceptableIdentifierCharacter(lexer.inputReader.currentCharacter)) {
    identifierBuilder.push(lexer.inputReader.currentCharacter!);
  }
  return identifierBuilder.join('');
}

export function isAcceptableIdentifierStart(char: string): boolean {
  return ALPHABET.includes(char) || char === '_';
}

export function isAcceptableIdentifierCharacter(char: string | null): boolean {
  if (char === null) {
    return false;
  }
  return isAcceptableIdentifierStart(char) || DIGITS.includes(char) || char === '-';
}

File: tools-jvm/interpreters/languages/java-lang-common/src/main/kotlin/interpreters/javalang/common/lexer/readIdentifier.kt
package interpreters.javalang.common.lexer

internal fun Lexer.readIdentifier(): String {
    val startIndex = currentIndex
    if (!currentCharacter.isAcceptableIdentifierStart()) {
        throw Error("You should not attempt to read an identifier which doesn't start with '_' or a letter")
    }
    while (currentCharacter.isAcceptableIdentifierNonStart()) {
        readNextCharacter()
    }
    return input.slice(startIndex..<currentIndex)
}
